shell/platform/common/text_editing_delta.h

```diff
@@ -0,0 +1,47 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_COMMON_TEXT_EDITING_DELTA_H_
+#define FLUTTER_SHELL_PLATFORM_COMMON_TEXT_EDITING_DELTA_H_
+
+#include <string>
+
+#include "flutter/shell/platform/common/text_range.h"
+
+namespace flutter {
+
+class TextEditingDelta {
+ public:
+  TextEditingDelta(std::string textBeforeChange,
+                   TextRange range,
+                   std::string text);
+  virtual ~TextEditingDelta();
+
+  std::string oldText() const { return oldText_; }
+  std::string deltaText() const { return deltaText_; }
+  int deltaStart() const { return deltaStart_; }
+  int deltaEnd() const { return deltaEnd_; }
+
+  TextEditingDelta(std::string text);
```

Comment:
Also can this be a struct since the deltas are short-lived and immutable for the most part? Not familiar with C++ not sure what the best practice here is.

Reply:
@cbracken Can you advise on these two points?

 1. I've made the class into a struct for now.  I don't know if it makes a difference in this case though.  I'm interested to know if there are best practices to follow.

 2. About UTF8/16.  It seems like Mac uses text_input_model, which internally stores the text in UTF16, but converts to UTF8 when accessed in GetText ([code](https://github.com/flutter/engine/blob/387563aecb9ff7ace912fdf7bc297700f3913641/shell/platform/common/text_input_model.cc#L292-L296)).  iOS stores the text in NSMutableString ([code](https://github.com/flutter/engine/blob/387563aecb9ff7ace912fdf7bc297700f3913641/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h#L60)), which is UTF16.  My TextEditingDelta class in this PR stores the text in a UTF8 string.  We'd like to minimize conversions, maybe by letting TextEditingDelta take any type by using a template?

Reply:
1. For structs vs classes, take a look at https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes. In C++ they are almost identical. In this case we could go with a struct, but I don't think using a class is particularly bad as well.
2. I'm not sure where this class gets instantiated, but depending on how much control we have we can either go with [u16string](https://www.cplusplus.com/reference/string/u16string/) or taking in a [basic_string](https://www.cplusplus.com/reference/string/basic_string/) which would cover both utf-8, 16.

Reply:
Apologies for missing this earlier. Agreed on point 1.

Point 2 will depend on how the string is encoded. If you're sending a UTF-8 string, use `std::string`. If the string is UTF-16, then use `std::u16string`. You'll notice in our Windows APIs that we make extensive use of `std::wstring` which is a string of `wchar_t`. In general, `std::wstring` and `std::u16string` are _NOT_ the same, but on Windows, they are.

In other places in the code that live across multiple platforms, you'll see that we provide multiple setters/getters. An example is [TextInputModel](https://github.com/flutter/engine/blob/main/shell/platform/common/text_input_model.h). In that case, we store the text as UTF-16 and provide setters for UTF-8 and UTF-16.

Reply:
+1 to what @cbracken said

Reply:
Thanks for the insights!

Regarding encoding, if you take a `NSString` and convert it to UTF-8 to store in a `TextEditingDelta`, when you send the `TextEditingDelta` to the framework the string needs to be convert back to `NSString` (I think?), and then `NSJSONSerialization` converts it to JSON (utf-8), unless `NSString` is smart enough to recognize the string is immutable and only does UTF-8 => UTF-16 conversion lazily.

Reply:
> unless NSString is smart enough to recognize the string is immutable and only does UTF-8 => UTF-16 conversion lazily. 

NSString has a fairly complex internal representation at least last I remember. For short ASCII-ish strings, it uses a [tagged pointer representation](https://mikeash.com/pyblog/friday-qa-2015-07-31-tagged-pointer-strings.html), for longer strings, it allocates a buffer and stores as a UTF-16 representation.

Reply:
Is there any action to take here on the utf8/16 topic then?  I could make TextEditingDelta store the text as std::u16string, but I'm unclear if that would better, worse, or the same in terms of number of conversions including both iOS and Mac.

Reply:
Given that this is a private, embedder-internal representation, I'd lean towards using storing as `std::u16string` and supporting both `std::string` and `std::u16string` setters/constructor. 

On Windows, that will almost certainly result in less conversions and it may also be the case with NSString, which, from memory, uses a UTF-16 representation for longer strings where it isn't using a tagged pointer representation.

Reply:
That's right, we can change this in the future without breaking the framework.

Sounds good to me, I'll port this over to store the text in TextEditingDelta as std::u16string.  I'll do it in one commit so it's easy to review.  Thanks for the clarification.